### PR TITLE
Use NYC for code coverage and compile typescript on fly during test execution

### LINKED ISSUE #307: /* istanbul ignore next */ is not working
because tsconfig.json now has `removeComments: true` istanbul does not get ignore next comment.
we need to remove it from main tsconfig.json and add another npm script that will run compile with removeComments.

### PR DESCRIPTION
**Pull request by soda0289**
InstabulJS code coverage tool has been renamed to NYC. This commit
upgrades istanbul and adds a basic confiugration for code coverage.
It also removes the need to compile typescript before testing or
checking code coverage by using ts-node to compile the typescript
 as needed.

The graphql-server-express required a test to be changed since
it was uploading the test file it self, which changes during
compilation.

Also koa is not compatible with node.js v4 so I had to add a check that
disables that test suite when run in that version.

TODO:

- [ ] Update CHANGELOG.md with your change (include reference to issue & this PR)
- [ ] Make sure all of the significant new logic is covered by tests
- [ ] Rebase your changes on master so that they can be merged easily
- [ ] Make sure all tests and linter rules pass

Closes #307

### CODE REVIEW

**Comment by soda0289** [comment]
Not sure why it shows up as 0% on travis not able to produce those results. It should fail if below 85%

File: package.json [hunk 1]
```diff
@@ -10,15 +10,13 @@
     "lint": "tslint ./packages/**/src/**/*.ts --exclude \"./packages/**/node_modules/**/*.ts\"",
     "prebootstrap": "npm install",
     "postinstall": "lerna bootstrap",
-    "pretest": "npm run compile",
```
Comment (DxCx) [inline package.json]: why this step is removed?

File: package.json [hunk 2]
```diff
@@ -10,15 +10,13 @@
...
     "posttest": "npm run lint",
-    "testonly": "mocha --reporter spec --full-trace --timeout 5000 ./test/tests.js",
-    "coverage": "istanbul cover -x \"*.test.js\" _mocha -- --timeout 5000 --full-trace --reporter dot ./test/tests.js",
+    "testonly": "mocha --reporter spec --full-trace --timeout 5000 --compilers ts:ts-node/register ./test/tests.js",
+    "coverage": "nyc _mocha --full-trace --timeout 5000 ./test/tests.js",
     "pretravis": "npm run compile",
-    "travis": "istanbul cover -x \"*.test.js\" _mocha -- --timeout 5000 --full-trace ./test/tests.js",
+    "travis": "npm run coverage",
     "posttravis": "npm run lint",
-    "postcoverage": "remap-istanbul --input coverage/coverage.raw.json --type lcovonly --output coverage/lcov.info",
```
Comment (DxCx) [inline package.json]: so nyc has built it remap?

File: test/tests.js [hunk 3]
```diff
@@ -1,11 +1,16 @@
...
-require('../packages/graphql-server-koa/dist/koaApollo.test');
-require('../packages/graphql-server-restify/dist/restifyApollo.test');
-require('../packages/graphql-server-lambda/dist/lambdaApollo.test');
-require('../packages/graphql-server-express/dist/apolloServerHttp.test');
+require('../packages/graphql-server-core/src/runQuery.test');
+require('../packages/graphql-server-module-operation-store/src/operationStore.test');
+require('../packages/graphql-server-express/src/expressApollo.test');
+require('../packages/graphql-server-express/src/connectApollo.test');
+require('../packages/graphql-server-hapi/src/hapiApollo.test');
+if (NODE_MAJOR_VERSION >= 6) {
```
Comment (DxCx) [inline test/tests.js]: Please open an issue about that, and then add a comment pointing to that issue, it should be fixed as well.

**Comment by soda0289** [comment]
I have figured out why it is showing 0%. I first must enable instrumentation with out the compiled dist directories, not sure why. Then disabling instrumentation and compiling with tsc works. Going to read thorough some source code and docs to figure out a clean solution

File: package.json (hunk 1 above)
Comment (soda0289) [inline package.json]: Compilation for the tests and code coverage is done on the fly. We can use ts-node to compile typescript files when we require them. This removes the need to precompile code to run tests

File: package.json (hunk 2 above)
Comment (soda0289) [inline package.json]: yes it can read source maps and require in ts-node for us.

File: test/tests.js (hunk 3 above)
Comment (soda0289) [inline test/tests.js]: Sure thing

**Comment by soda0289** [comment]
This will take some time as there are bugs in nyc that prevent this working with the multi package setup. Will break into smaller PRs for nodejs v4 fix and updating tests to use ts-node

**Comment by DxCx** [comment]
alright, thanks 👍

**Pull request closed without merging**